File: lib/models/writing.dart
/// Represents a single writing/poem/note
class Writing {
  final String id;
  String title;
  String body;
  String footer;
  final DateTime createdAt;
  DateTime updatedAt;
  bool isSynced;
  bool isBold;
  String textAlign; // 'left', 'center', 'right'
  DateTime? deletedAt; // Soft-delete timestamp (null = not deleted)

  Writing({
    required this.id,
    required this.title,
    this.body = '',
    this.footer = '',
    required this.createdAt,
    required this.updatedAt,
    this.isSynced = false,
    this.isBold = false,
    this.textAlign = 'left',
    this.deletedAt,
  });

  /// Whether this writing is soft-deleted
  bool get isDeleted => deletedAt != null;

  /// Create a new writing with generated ID and timestamps
  factory Writing.create({
    required String title,
    String body = '',
    String footer = '',
    bool isBold = false,
    String textAlign = 'left',
  }) {
    final now = DateTime.now();
    return Writing(
      id: DateTime.now().millisecondsSinceEpoch.toString(),
      title: title,
      body: body,
      footer: footer,
      createdAt: now,
      updatedAt: now,
      isSynced: false,
      isBold: isBold,
      textAlign: textAlign,
      deletedAt: null,
    );
  }

  /// Create a copy with updated fields
  Writing copyWith({
    String? title,
    String? body,
    String? footer,
    DateTime? updatedAt,
    bool? isSynced,
    bool? isBold,
    String? textAlign,
    DateTime? deletedAt,
    bool clearDeletedAt = false,
  }) {
    return Writing(
      id: id,
      title: title ?? this.title,
      body: body ?? this.body,
      footer: footer ?? this.footer,
      createdAt: createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isSynced: isSynced ?? this.isSynced,
      isBold: isBold ?? this.isBold,
      textAlign: textAlign ?? this.textAlign,
      deletedAt: clearDeletedAt ? null : (deletedAt ?? this.deletedAt),
    );
  }

  /// Get a preview of the body (first 50 characters)
  /// Strips HTML tags if present
  String get bodyPreview {
    if (body.isEmpty) return '';
    
    // Strip HTML tags for preview
    String preview = body
        .replaceAll(RegExp(r'<[^>]*>'), '')
        .replaceAll('&lt;', '<')
        .replaceAll('&gt;', '>')
        .replaceAll('&quot;', '"')
        .replaceAll('&#39;', "'")
        .replaceAll('&amp;', '&')
        .trim();
    
    return preview.length > 50 ? '${preview.substring(0, 50)}...' : preview;
  }
}

File: src/types/writing.ts
// Writing types matching Flutter app

export type WritingType = 'siir' | 'yazi' | 'diger';

export interface Writing {
  id: string;
  title: string;
  body: string;
  footer: string;
  createdAt: string; // ISO timestamp
  updatedAt: string;
  isSynced: boolean;
  isBold: boolean;
  textAlign: 'left' | 'center' | 'right';
  deletedAt?: string; // Soft-delete timestamp
  type: WritingType;
  stars?: number; // 0-5 star rating (0 = no rating)
}

export interface WritingMetadata {
  id: string;
  title: string;
  preview: string; // First 100 chars of body
  createdAt: string;
  updatedAt: string;
  isSynced: boolean;
  deletedAt?: string;
  type: WritingType;
  stars?: number; // 0-5 star rating (0 = no rating)
}

export interface MetadataIndex {
  version: number;
  lastSyncTime?: string;
  writings: WritingMetadata[];
}

// Current metadata schema version - increment when structure changes
export const METADATA_INDEX_VERSION = 3;

// Helper functions
export const writingTypeDisplayName: Record<WritingType, string> = {
  siir: 'Şiir',
  yazi: 'Yazı',
  diger: 'Diğer',
};

export function createWriting(params: {
  title?: string;
  body?: string;
  footer?: string;
  type?: WritingType;
  isBold?: boolean;
  textAlign?: 'left' | 'center' | 'right';
  stars?: number;
}): Writing {
  const now = new Date().toISOString();
  return {
    id: Date.now().toString(),
    title: params.title ?? '',
    body: params.body ?? '',
    footer: params.footer ?? '',
    createdAt: now,
    updatedAt: now,
    isSynced: false,
    isBold: params.isBold ?? false,
    textAlign: params.textAlign ?? 'left',
    type: params.type ?? 'siir',
    stars: params.stars ?? 0,
  };
}

export function generatePreview(body: string): string {
  if (!body) return '';
  
  // Strip HTML tags and clean up
  const preview = body
    .replace(/<[^>]*>/g, '')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&amp;/g, '&')
    .replace(/\s+/g, ' ')
    .trim();
  
  return preview.length > 100 ? `${preview.substring(0, 100)}...` : preview;
}

export function metadataFromWriting(writing: Writing): WritingMetadata {
  return {
    id: writing.id,
    title: writing.title,
    preview: generatePreview(writing.body),
    createdAt: writing.createdAt,
    updatedAt: writing.updatedAt,
    isSynced: writing.isSynced,
    deletedAt: writing.deletedAt,
    type: writing.type,
    stars: writing.stars ?? 0,
  };
}

export function isDeleted(item: Writing | WritingMetadata): boolean {
  return item.deletedAt !== undefined && item.deletedAt !== null;
}

/**
 * Check if a writing is valid for saving.
 * A writing must have either a non-empty title OR non-empty body (content) to be saved.
 */
export function isValidForSave(writing: Writing): boolean {
  return writing.title.trim().length > 0 || writing.body.trim().length > 0;
}
